superfluous_disable_command not triggering for custom rule

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

`superfluous_disable_command` is not triggering for custom rules. I don't know if this has always been the behavior or if it's a recent regression.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint file.swift           
Linting Swift files at paths file.swift
Linting 'file.swift' (1/1)
/Users/marcelo.fabri/SwiftLint/test/file.swift:1:5: warning: forbidden Violation: This is forbidden (forbidden)
Done linting! Found 1 violation, 0 serious in 1 file.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.50.3
* Installation method used (Homebrew, CocoaPods, building from source, etc)? brew
* Paste your configuration file:

```yml
custom_rules:
  forbidden:
    regex: 'FORBIDDEN'
    message: "This is forbidden"
    severity: warning
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? No
* Which Xcode version are you using (check `xcodebuild -version`)? 14.0.1
* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```swift
let FORBIDDEN = 1
// swiftlint:disable:next forbidden
let ALLOWED = 2
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -68,6 +68,12 @@
   [jszumski](https://github.com/jszumski)
   [#5246](https://github.com/realm/SwiftLint/pull/5246)
 
+* `superfluous_disable_command` violations are now triggered for
+  custom rules.  
+  [Marcelo Fabri](https://github.com/marcelofabri)
+  [Martin Redington](https://github.com/mildm8nnered)
+  [#4754](https://github.com/realm/SwiftLint/issues/4754)
+
 ## 0.53.0: Laundry List
 
 #### Breaking
@@ -730,12 +736,6 @@
 * Updated JUnit reporter to output error count and warning count.  
   [patricks](https://github.com/patricks)
   [#4725](https://github.com/realm/SwiftLint/pull/4725)
-
-* `superfluous_disable_command` violations are now triggered for
-  custom rules.  
-  [Marcelo Fabri](https://github.com/marcelofabri)
-  [Martin Redington](https://github.com/mildm8nnered)
-  [#4754](https://github.com/realm/SwiftLint/issues/4754)
 
 * Fix correction on `lower_acl_than_parent` rule for `open` declarations.  
   [Marcelo Fabri](https://github.com/marcelofabri)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -731,6 +731,12 @@
   [patricks](https://github.com/patricks)
   [#4725](https://github.com/realm/SwiftLint/pull/4725)
 
+* `superfluous_disable_command` violations are now triggered for
+  custom rules.  
+  [Marcelo Fabri](https://github.com/marcelofabri)
+  [Martin Redington](https://github.com/mildm8nnered)
+  [#4754](https://github.com/realm/SwiftLint/issues/4754)
+
 * Fix correction on `lower_acl_than_parent` rule for `open` declarations.  
   [Marcelo Fabri](https://github.com/marcelofabri)
   [#4753](https://github.com/realm/SwiftLint/issues/4753)

--- a/Source/SwiftLintCore/Models/Linter.swift
+++ b/Source/SwiftLintCore/Models/Linter.swift
@@ -16,42 +16,57 @@ private struct LintResult {
 }
 
 private extension Rule {
-    static func superfluousDisableCommandViolations(regions: [Region],
-                                                    superfluousDisableCommandRule: SuperfluousDisableCommandRule?,
-                                                    allViolations: [StyleViolation]) -> [StyleViolation] {
+    func superfluousDisableCommandViolations(regions: [Region],
+                                             superfluousDisableCommandRule: SuperfluousDisableCommandRule?,
+                                             allViolations: [StyleViolation]) -> [StyleViolation] {
         guard regions.isNotEmpty, let superfluousDisableCommandRule else {
             return []
         }
 
-        let regionsDisablingCurrentRule = regions.filter { region in
-            return region.isRuleDisabled(self.init())
-        }
         let regionsDisablingSuperfluousDisableRule = regions.filter { region in
             return region.isRuleDisabled(superfluousDisableCommandRule)
         }
 
-        return regionsDisablingCurrentRule.compactMap { region -> StyleViolation? in
-            let isSuperfluousRuleDisabled = regionsDisablingSuperfluousDisableRule.contains {
-                $0.contains(region.start)
+        let regionsWithIdentifiers: [(String, [Region])] = {
+            if let customRules = self as? CustomRules {
+                return customRules.configuration.customRuleConfigurations.map { configuration in
+                    let regionsDisablingCurrentRule = regions.filter { region in
+                        return region.isRuleDisabled(customRuleIdentifier: configuration.identifier)
+                    }
+                    return (configuration.identifier, regionsDisablingCurrentRule)
+                }
+            } else {
+                let regionsDisablingCurrentRule = regions.filter { region in
+                    return region.isRuleDisabled(self)
+                }
+                return [(Self.description.identifier, regionsDisablingCurrentRule)]
             }
+        }()
 
-            guard !isSuperfluousRuleDisabled else {
-                return nil
-            }
+        return regionsWithIdentifiers.flatMap { ruleIdentifier, regionsDisablingCurrentRule in
+            regionsDisablingCurrentRule.compactMap { region -> StyleViolation? in
+                let isSuperfluousRuleDisabled = regionsDisablingSuperfluousDisableRule.contains {
+                    $0.contains(region.start)
+                }
 
-            let noViolationsInDisabledRegion = !allViolations.contains { violation in
-                return region.contains(violation.location)
-            }
-            guard noViolationsInDisabledRegion else {
-                return nil
-            }
+                guard !isSuperfluousRuleDisabled else {
+                    return nil
+                }
 
-            return StyleViolation(
-                ruleDescription: type(of: superfluousDisableCommandRule).description,
-                severity: superfluousDisableCommandRule.configuration.severity,
-                location: region.start,
-                reason: superfluousDisableCommandRule.reason(for: self)
-            )
+                let noViolationsInDisabledRegion = !allViolations.contains { violation in
+                    return region.contains(violation.location)
+                }
+                guard noViolationsInDisabledRegion else {
+                    return nil
+                }
+
+                return StyleViolation(
+                    ruleDescription: type(of: superfluousDisableCommandRule).description,
+                    severity: superfluousDisableCommandRule.configuration.severity,
+                    location: region.start,
+                    reason: superfluousDisableCommandRule.reason(forRuleIdentifier: ruleIdentifier)
+                )
+            }
         }
     }
 
@@ -91,12 +106,19 @@ private extension Rule {
             return region?.isRuleEnabled(self) ?? true
         }
 
+        let customRulesIDs: [String] = {
+            guard let customRules = self as? CustomRules else {
+                return []
+            }
+            return customRules.configuration.customRuleConfigurations.map(\.identifier)
+        }()
         let ruleIDs = Self.description.allIdentifiers +
+            customRulesIDs +
             (superfluousDisableCommandRule.map({ type(of: $0) })?.description.allIdentifiers ?? []) +
             [RuleIdentifier.all.stringRepresentation]
         let ruleIdentifiers = Set(ruleIDs.map { RuleIdentifier($0) })
 
-        let superfluousDisableCommandViolations = Self.superfluousDisableCommandViolations(
+        let superfluousDisableCommandViolations = superfluousDisableCommandViolations(
             regions: regions.count > 1 ? file.regions(restrictingRuleIdentifiers: ruleIdentifiers) : regions,
             superfluousDisableCommandRule: superfluousDisableCommandRule,
             allViolations: violations

--- a/Source/SwiftLintCore/Models/Linter.swift
+++ b/Source/SwiftLintCore/Models/Linter.swift
@@ -54,7 +54,7 @@ private extension Rule {
                 }
 
                 let noViolationsInDisabledRegion = !allViolations.contains { violation in
-                    return region.contains(violation.location)
+                    return region.contains(violation.location) && violation.ruleIdentifier == ruleIdentifier
                 }
                 guard noViolationsInDisabledRegion else {
                     return nil

--- a/Source/SwiftLintCore/Rules/CustomRules.swift
+++ b/Source/SwiftLintCore/Rules/CustomRules.swift
@@ -90,7 +90,7 @@ struct CustomRules: Rule, CacheDescriptionProvider {
     }
 }
 
-private extension Region {
+extension Region {
     func isRuleDisabled(customRuleIdentifier: String) -> Bool {
         return disabledRuleIdentifiers.contains(RuleIdentifier(customRuleIdentifier))
     }

--- a/Source/SwiftLintCore/Rules/SuperfluousDisableCommandRule.swift
+++ b/Source/SwiftLintCore/Rules/SuperfluousDisableCommandRule.swift
@@ -35,9 +35,9 @@ public struct SuperfluousDisableCommandRule: SourceKitFreeRule {
         return []
     }
 
-    func reason(for rule: (some Rule).Type) -> String {
+    func reason(forRuleIdentifier ruleIdentifier: String) -> String {
         """
-        SwiftLint rule '\(rule.description.identifier)' did not trigger a violation in the disabled region; \
+        SwiftLint rule '\(ruleIdentifier)' did not trigger a violation in the disabled region; \
         remove the disable command
         """
     }

--- a/Tests/SwiftLintFrameworkTests/CustomRulesTests.swift
+++ b/Tests/SwiftLintFrameworkTests/CustomRulesTests.swift
@@ -201,7 +201,7 @@ class CustomRulesTests: SwiftLintTestCase {
         })
     }
 
-    func testSuperfluousDisableCommandWithMultipleCustomRules()  throws{
+    func testSuperfluousDisableCommandWithMultipleCustomRules() throws {
         let customRulesConfiguration: [String: Any] = [
             "custom1": [
                 "regex": "pattern",

--- a/Tests/SwiftLintFrameworkTests/CustomRulesTests.swift
+++ b/Tests/SwiftLintFrameworkTests/CustomRulesTests.swift
@@ -179,7 +179,7 @@ class CustomRulesTests: SwiftLintTestCase {
         XCTAssertEqual(violations[0].location.character, 6)
     }
 
-    func testSuperfluousDisableCommandWithCustomRules() {
+    func testSuperfluousDisableCommandWithCustomRules() throws {
         let customRulesConfiguration: [String: Any] = [
             "custom1": [
                 "regex": "pattern",
@@ -191,7 +191,7 @@ class CustomRulesTests: SwiftLintTestCase {
             "// swiftlint:disable custom1\n",
             configuration: customRulesConfiguration
         ).skipWrappingInCommentTest()
-        let configuration = try! Configuration(dict: ["custom_rules": customRulesConfiguration])
+        let configuration = try Configuration(dict: ["custom_rules": customRulesConfiguration])
         let violations = violations(example, config: configuration)
 
         XCTAssertEqual(violations.count, 1)
@@ -201,7 +201,7 @@ class CustomRulesTests: SwiftLintTestCase {
         })
     }
 
-    func testSuperfluousDisableCommandWithMultipleCustomRules() {
+    func testSuperfluousDisableCommandWithMultipleCustomRules()  throws{
         let customRulesConfiguration: [String: Any] = [
             "custom1": [
                 "regex": "pattern",
@@ -225,7 +225,7 @@ class CustomRulesTests: SwiftLintTestCase {
             """,
             configuration: customRulesConfiguration
         ).skipWrappingInCommentTest()
-        let configuration = try! Configuration(dict: ["custom_rules": customRulesConfiguration])
+        let configuration = try Configuration(dict: ["custom_rules": customRulesConfiguration])
         let violations = violations(example, config: configuration)
 
         XCTAssertEqual(violations.count, 3)

--- a/Tests/SwiftLintFrameworkTests/CustomRulesTests.swift
+++ b/Tests/SwiftLintFrameworkTests/CustomRulesTests.swift
@@ -191,7 +191,7 @@ class CustomRulesTests: SwiftLintTestCase {
             "// swiftlint:disable custom1\n",
             configuration: customRulesConfiguration
         ).skipWrappingInCommentTest()
-        let configuration = try Configuration(dict: ["custom_rules": customRulesConfiguration])
+        let configuration = try XCTUnwrap(makeConfig(["custom_rules": customRulesConfiguration], "custom_rules"))
         let violations = violations(example, config: configuration)
 
         XCTAssertEqual(violations.count, 1)
@@ -225,7 +225,7 @@ class CustomRulesTests: SwiftLintTestCase {
             """,
             configuration: customRulesConfiguration
         ).skipWrappingInCommentTest()
-        let configuration = try Configuration(dict: ["custom_rules": customRulesConfiguration])
+        let configuration = try XCTUnwrap(makeConfig(["custom_rules": customRulesConfiguration], "custom_rules"))
         let violations = violations(example, config: configuration)
 
         XCTAssertEqual(violations.count, 3)

--- a/Tests/SwiftLintFrameworkTests/CustomRulesTests.swift
+++ b/Tests/SwiftLintFrameworkTests/CustomRulesTests.swift
@@ -180,7 +180,7 @@ class CustomRulesTests: SwiftLintTestCase {
     }
 
     func testSuperfluousDisableCommandWithCustomRules() {
-        let customRulesConfiguration: [String:Any] = [
+        let customRulesConfiguration: [String: Any] = [
             "custom1": [
                 "regex": "pattern",
                 "match_kinds": "comment"
@@ -202,7 +202,7 @@ class CustomRulesTests: SwiftLintTestCase {
     }
 
     func testSuperfluousDisableCommandWithMultipleCustomRules() {
-        let customRulesConfiguration: [String:Any] = [
+        let customRulesConfiguration: [String: Any] = [
             "custom1": [
                 "regex": "pattern",
                 "match_kinds": "comment"

--- a/Tests/SwiftLintFrameworkTests/CustomRulesTests.swift
+++ b/Tests/SwiftLintFrameworkTests/CustomRulesTests.swift
@@ -179,6 +179,66 @@ class CustomRulesTests: SwiftLintTestCase {
         XCTAssertEqual(violations[0].location.character, 6)
     }
 
+    func testSuperfluousDisableCommandWithCustomRules() {
+        let customRulesConfiguration: [String:Any] = [
+            "custom1": [
+                "regex": "pattern",
+                "match_kinds": "comment"
+            ]
+        ]
+
+        let example = Example(
+            "// swiftlint:disable custom1\n",
+            configuration: customRulesConfiguration
+        ).skipWrappingInCommentTest()
+        let configuration = try! Configuration(dict: ["custom_rules": customRulesConfiguration])
+        let violations = violations(example, config: configuration)
+
+        XCTAssertEqual(violations.count, 1)
+        XCTAssertTrue(violations.allSatisfy { $0.ruleIdentifier == "superfluous_disable_command" })
+        XCTAssertTrue(violations.contains { violation in
+            violation.description.contains("SwiftLint rule 'custom1' did not trigger a violation")
+        })
+    }
+
+    func testSuperfluousDisableCommandWithMultipleCustomRules() {
+        let customRulesConfiguration: [String:Any] = [
+            "custom1": [
+                "regex": "pattern",
+                "match_kinds": "comment"
+            ],
+            "custom2": [
+                "regex": "10",
+                "match_kinds": "number"
+            ],
+            "custom3": [
+                "regex": "100",
+                "match_kinds": "number"
+            ]
+        ]
+
+        let example = Example(
+            """
+            // swiftlint:disable custom1 custom3
+            return 10
+
+            """,
+            configuration: customRulesConfiguration
+        ).skipWrappingInCommentTest()
+        let configuration = try! Configuration(dict: ["custom_rules": customRulesConfiguration])
+        let violations = violations(example, config: configuration)
+
+        XCTAssertEqual(violations.count, 3)
+        XCTAssertEqual(violations.filter { $0.ruleIdentifier == "superfluous_disable_command" }.count, 2)
+        XCTAssertEqual(violations.filter { $0.ruleIdentifier == "custom2" }.count, 1)
+        XCTAssertTrue(violations.contains { violation in
+            violation.description.contains("SwiftLint rule 'custom1' did not trigger a violation")
+        })
+        XCTAssertTrue(violations.contains { violation in
+            violation.description.contains("SwiftLint rule 'custom3' did not trigger a violation")
+        })
+    }
+
     private func getCustomRules(_ extraConfig: [String: Any] = [:]) -> (Configuration, CustomRules) {
         var config: [String: Any] = ["regex": "pattern",
                                      "match_kinds": "comment"]

